ExoCom: Show response time of service

martinjaime opened the pull request:
@kevgo @helenmedrano 

I had not noticed I had been unassigned, but I just finished this issue. This pull request is paired with [/Originate/exosphere-sdk#139](https://github.com/Originate/exosphere-sdk/pull/139)

Addresses [/Originate/exosphere-sdk#30](https://github.com/Originate/exosphere-sdk/issues/30)

kevgo commented on features/broadcasting-messages.feature:
```diff
@@ -25,5 +25,5 @@ Feature: Broadcasting messages
   Scenario: broadcasting a reply
     When the "users" service sends "created-user" in reply to "111"
-    Then ExoCom signals "users  --[ created-user ]->  web"
+    Then ExoCom signals "users  --[ created-user ]->  web  response time: 123456789ns"
```
Let's just print:

```
users  --[ created-user ]->  web   (12ms)

kevgo commented on src/listener-subsystem/zmq-listener.ls:
```diff
@@ -59,8 +59,10 @@ class ZMQListener extends EventEmitter
+    response-time = req.response-time
     id = req.id
-    {sender, name, response-to, payload, id}
+    {sender, name, payload, id, response-to, timestamp, response-time}
```
There are a number of entries here now, let's keep them sorted alphabetically.

kevgo commented:
Nicely done! Let's see how this looks like with the shorter formatting.

martinjaime commented:
@kevgo All comments addressed. 
![image](https://cloud.githubusercontent.com/assets/10568301/18180509/0022e838-703c-11e6-9fa1-d93fd703754e.png)

kevgo commented on features/support/api-world.ls:
```diff
@@ -70,16 +70,17 @@ ApiWorld = !->
       | 5  =>  [_, original-name, _, translated-name, _] = message-name-parts
       | _  =>  throw new Error "Unknown message name parts"
     wait-until (~> @last-sent-messages[0].name is translated-name), 1, ~>
+      [seconds, nanoseconds] = @last-sent-messages[0].timestamp
```
Let's make the `timestamp` field a single number, consisting nanoseconds. This is hopefully a bit shorter and more compatible with other languages.

kevgo commented on features/support/cli-world.ls:
```diff
@@ -81,7 +81,9 @@ CliWorld = !->
   @verify-exocom-signaled-string = (message, done) ->
-    @process.wait message, done
+    [...message-main, response-time-msg] = message.split '  '
+    response-time = response-time-msg.match(/-?\d+\.\d+/)?[0]
```
You aren't using the local variable `response-time` here. Can you remove this line?

kevgo commented on features/support/cli-world.ls:
```diff
@@ -81,7 +81,9 @@ CliWorld = !->
-    @process.wait message, done
+    [...message-main, response-time-msg] = message.split '  '
+    response-time = response-time-msg.match(/-?\d+\.\d+/)?[0]
+    @process.wait "#{message-main.join '  '}  ", done
```
I think overall this is a pretty brittle way of parsing this string. Let's try using a regular expression. How do you feel about something like this:

```ls
expected-string = message.match/(.*)  \(\d\.\d ms\)/[1]
@process.wait expected-string, done

kevgo commented:
I love how the colored output looks like right now!

kevgo commented on features/translating-messages.feature:
```diff
@@ -27,5 +27,5 @@ Feature: Translating messages
   Scenario: translating a reply
     When the "tweets" service sends "text-snippets.created" in reply to "111"
-    Then ExoCom signals "tweets  --[ text-snippets.created ]-[ tweets.created ]->  web"
+    Then ExoCom signals "tweets  --[ text-snippets.created ]-[ tweets.created ]->  web  response time: 123456789ns"
```
This is still the old format

kevgo commented on src/cli.ls:
```diff
@@ -49,10 +49,14 @@ run = ->
+        response-time = ''
+        if message.response-to
+          [seconds, nanoseconds] = process.hrtime(message.timestamp)
+          response-time = "  (#{(seconds * 1e3 + nanoseconds / 1e6).to-fixed 1} ms)"
```
This will become simpler when you switch to representing time as a single number of nanoseconds since the epoch.

kevgo replied:
No need to store a separate string here. Just add it to line 57

kevgo commented on src/listener-subsystem/http-listener.ls:
```diff
@@ -50,8 +50,8 @@ class HttpListener extends EventEmitter
-  _log: ({name, id, response-to}) ->
-    | response-to  =>  debug "received message '#{name}' with id '#{id}' in response to '#{response-to}'"
+  _log: ({name, id, response-to, timestamp}) ->
+    | response-to  =>  debug "received message '#{name}' with id '#{id}' in response to '#{response-to}' | response time: #{timestamp}"
```
still old format here

martinjaime commented:
@kevgo, I think this should be the last review. I could not find a module to give accurate nanosecond readings. `Date.now!` will give us the Unix timestamp we want with millisecond precision. We could also use `process.hrtime!` for nanosecond precision, but that will not give us a Unix timestamp. This means we would only be able to use the timestamp to measure response time.

kevgo commented:
Please use `process.hrtime`. When you give it the original timestamp as an argument, it gives you the diff, see https://nodejs.org/api/process.html#process_process_hrtime_time. That should be all you need here.

kevgo commented on features/translating-messages.feature:
```diff
@@ -26,6 +26,7 @@ Feature: Translating messages
-    Then ExoCom signals "tweets  --[ text-snippets.created ]-[ tweets.created ]->  web"
+    When the "web" service sends "tweets.create"
+    And the "tweets" service sends "text-snippets.created" in reply to "111"
+    Then ExoCom signals "tweets  --[ text-snippets.created ]-[ tweets.created ]->  web  (8 ms)"
```
Are you sure it's always 8 ms here?

kevgo replied:
Can you say `(XX ms)` here to make clear that this is a placeholder?

kevgo commented on src/exocom.ls:
```diff
@@ -47,13 +47,12 @@ class ExoCom extends EventEmitter
     message-data.original-name = message-data.name
     message-data.name = external-message-name
-    message-data.timestamp = seconds * 1e9 + nanos
+    message-data.timestamp = Date.now!
```
Please keep this as it was before. Before it was an integer timestamp with nanosecond precision. Which is what we want. Now it is a Date object with millisecond precision. We don't want that.

martinjaime replied:
Yes, I originally wanted to use `process.hrtime`, but switched when I found out that it was not Unix time. I will switch it, so we can have nanosecond precision.

kevgo commented on src/listener-subsystem/zmq-listener.ls:
```diff
@@ -59,8 +59,10 @@ class ZMQListener extends EventEmitter
+    response-time = req.response-time
     id = req.id
-    {sender, name, response-to, payload, id}
+    {id, name, payload, response-time, response-to, sender, timestamp}
```
Thanks for sorting this alphabetically!

kevgo commented on src/listener-subsystem/zmq-listener.ls:
```diff
@@ -59,8 +59,10 @@ class ZMQListener extends EventEmitter
     payload = req.payload
     response-to = req.response-to
+    timestamp = req.timestamp
+    response-time = req.response-time
```
The incoming message should not contain the response time. ExoCom should calculate this message on its own.

kevgo commented on src/listener-subsystem/http-listener.ls:
```diff
@@ -50,8 +50,8 @@ class HttpListener extends EventEmitter
-  _log: ({name, id, response-to}) ->
-    | response-to  =>  debug "received message '#{name}' with id '#{id}' in response to '#{response-to}'"
+  _log: ({name, id, response-to, timestamp}) ->
+    | response-to  =>  debug "received message '#{name}' with id '#{id}' in response to '#{response-to}'  (#{timestamp} ms)"
```
I don't think we need to add the timing to the debug messages. The debug statement is rendered with its own timestamp.

kevgo commented on src/listener-subsystem/zmq-listener.ls:
```diff
@@ -48,8 +48,8 @@ class ZMQListener extends EventEmitter
-  _log: ({name, id, response-to}) ->
-    | response-to  =>  debug "received message '#{name}' with id '#{id}' in response to '#{response-to}'"
+  _log: ({name, id, response-to, timestamp}) ->
+    | response-to  =>  debug "received message '#{name}' with id '#{id}' in response to '#{response-to}'  (#{timestamp} ms)"
```
Same here, this shouldn't be necessary.

kevgo commented on src/cli.ls:
```diff
@@ -49,10 +49,13 @@ run = ->
       for message in messages
+        response-time = ''
+        if message.response-to
+          response-time = "  (#{message.response-time} ms)"
```
Make sure this is rounded properly. It should have 2 digits after the comma, i.e. look like "1.23 ms"

martinjaime commented on src/message-sender/message-sender.ls:
```diff
@@ -44,13 +44,15 @@ class MessageSender extends EventEmitter
       timestamp: message-data.timestamp
     if message-data.response-to
       original-timestamp = @message-cache.get-original-timestamp request-data.id
+      request-data.response-time = Date.now! - original-timestamp
```
@kevgo, the response time is calculated here.

martinjaime commented:
@kevgo, I've moved all the response time logic from `message-sender` to the main `exocom` script. It turned out to be much more simpler. 

Also, added the `nanoseconds` npm module to do the simple conversion from the `process.hrtime` object into nanoseconds. It's simpler if we take the difference ourselves inline rather than calling the `process` API.

kevgo commented on package.json:
```diff
@@ -8,6 +8,7 @@
     "debug": "2.2.0",
     "docopt": "0.6.2",
     "express": "4.14.0",
+    "nanoseconds": "^0.1.0",
```
Remove the `^`. Also, run `npm config set save-exact true` so that NPM does this correct from now on.

kevgo commented on src/exocom.ls:
```diff
@@ -2,7 +2,9 @@ require! {
   './client-registry/client-registry' : ClientRegistry
   'events' : {EventEmitter}
   './listener-subsystem/listener-subsystem' : ListenerSubsystem
+  './message-cache/message-cache' : MessageCache
```
Can you sync this branch? I think these changes are unrelated.

martinjaime replied:
I had to move `message-cache` from `message-sender` to the main exocom script. So that is why this change is here. `message-cache` is unnecessary in `message-sender`. Should I create a new PR for that?

kevgo replied:
Ah, very cool! No, this is perfect.

kevgo commented on src/exocom.ls:
```diff
@@ -2,7 +2,9 @@ require! {
   './listener-subsystem/listener-subsystem' : ListenerSubsystem
+  './message-cache/message-cache' : MessageCache
   './message-sender/message-sender' : MessageSender
+  'nanoseconds' : Nanos
```
The convention is that identifiers with uppercase names are classes that need to be instantiated. Here you get a function back. I recommend naming it `nanoseconds`, similar to the module name.

kevgo commented on src/exocom.ls:
```diff
@@ -47,18 +50,23 @@ class ExoCom extends EventEmitter
     message-data.name = external-message-name
-    message-data.timestamp = seconds * 1e9 + nanos
-
+    message-data.timestamp = Nanos process.hrtime!
```
Great idea to use the nanoseconds library here!

kevgo commented on src/message-cache/message-cache.ls:
```diff
@@ -1,5 +1,6 @@
 require! {
   'wait' : {repeat}
+  'nanoseconds' : Nanos
```
same here, don't name this like a class

kevgo commented on src/message-cache/message-cache.ls:
```diff
@@ -9,8 +10,8 @@ class MessageCache
     repeat @cleanup-interval, ->
-      now = Date.now!
-      for id, timestamp of @cache when (timestamp - now) >= @cleanup-interval
+      now = Nanos process.hrtime!
```
No need to use nanosecond resolution here. You can keep it the old way. We just need to do regular cleanups at some point in time.  :)

kevgo replied:
Ah, you compare the timestamp in the messages to the current time stamp. Makes sense!

martinjaime commented:
@kevgo, I addressed and replied to your comments

kevgo commented on src/cli.ls:
```diff
@@ -49,10 +49,14 @@ run = ->
       for message in messages
+        response-time = ''
+        if message.response-to
+          response-time-ms = message.response-time * 1e-6
```
Inline this variable, it is used only once
